<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of adapterTemplate</title>
  <meta name="keywords" content="adapterTemplate">
  <meta name="description" content="">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../../index.html">Home</a> &gt;  <a href="#">src</a> &gt; <a href="#">geckomat</a> &gt; <a href="index.html">model_adapter</a> &gt; adapterTemplate.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../../index.html"><img alt="<" border="0" src="../../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for src\geckomat\model_adapter&nbsp;<img alt=">" border="0" src="../../../right.png"></a></td></tr></table>-->

<h1>adapterTemplate
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="box"><strong></strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="fragment"><pre class="comment"></pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../../matlabicon.gif)">
<li><a href="ModelAdapter.html" class="code" title="">ModelAdapter</a>	Abstract Base class for adapters for different species</li></ul>
This function is called by:
<ul style="list-style-image:url(../../../matlabicon.gif)">
</ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<ul style="list-style-image:url(../../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function obj = KEY_CLASSNAME()</a></li><li><a href="#_sub2" class="code">function [spont,spontRxnNames] = getSpontaneousReactions(obj,model)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="fragment"><pre>0001 classdef <a href="#_sub1" class="code" title="subfunction obj = KEY_CLASSNAME()">KEY_CLASSNAME</a> &lt; <a href="ModelAdapter.html" class="code" title="">ModelAdapter</a>
0002     methods
0003         <a name="_sub0" href="#_subfunctions" class="code">function obj = KEY_CLASSNAME()</a>
0004             <span class="comment">% Set initial values of the obj.params - they can be changed by the user</span>
0005             
0006             <span class="comment">% Directory where all model-specific files and scripts are kept.</span>
0007             <span class="comment">% Is assumed to follow the GECKO-defined folder structure.</span>
0008             obj.params.path = fullfile(<span class="string">'KEY_PATH'</span>, <span class="string">'KEY_NAME'</span>);
0009 
0010             <span class="comment">% Path to the conventional GEM that this ecModel will be based on.</span>
0011             obj.params.convGEM = fullfile(obj.params.path,<span class="string">'models'</span>,<span class="string">'yourModel.xml'</span>);
0012 
0013             <span class="comment">% Average enzyme saturation factor</span>
0014             obj.params.sigma = 0.5;
0015 
0016             <span class="comment">% Total protein content in the cell [g protein/gDw]</span>
0017             obj.params.Ptot = 0.5;
0018 
0019             <span class="comment">% Fraction of enzymes in the model [g enzyme/g protein]</span>
0020             obj.params.f = 0.5;
0021             
0022             <span class="comment">% Growth rate the model should be able to reach when not</span>
0023             <span class="comment">% constraint by nutrient uptake (e.g. max growth rate) [1/h]</span>
0024             obj.params.gR_exp = 0.41;
0025 
0026             <span class="comment">% Provide your organism scientific name</span>
0027             obj.params.org_name = <span class="string">'genus species'</span>;
0028             
0029             <span class="comment">% Taxonomic identifier for Complex Portal</span>
0030             obj.params.complex.taxonomicID = [];
0031 
0032             <span class="comment">% Provide your organism KEGG ID, selected at</span>
0033             <span class="comment">% https://www.genome.jp/kegg/catalog/org_list.html</span>
0034             obj.params.kegg.ID = <span class="string">'sce'</span>;
0035             <span class="comment">% Field for KEGG gene identifier; should match the gene</span>
0036             <span class="comment">% identifiers used in the model. With 'kegg', it takes the</span>
0037             <span class="comment">% default KEGG Entry identifier (for example YER023W here:</span>
0038             <span class="comment">% https://www.genome.jp/dbget-bin/www_bget?sce:YER023W).</span>
0039             <span class="comment">% Alternatively, gene identifiers from the &quot;Other DBs&quot; section</span>
0040             <span class="comment">% of the KEGG page can be selected. For example &quot;NCBI-GeneID&quot;,</span>
0041             <span class="comment">% &quot;UniProt&quot;, or &quot;Ensembl&quot;. Not all DB entries are available for</span>
0042             <span class="comment">% all organisms and/or genes.</span>
0043             obj.params.kegg.geneID = <span class="string">'kegg'</span>;
0044 
0045             <span class="comment">% Provide what identifier should be used to query UniProt.</span>
0046             <span class="comment">% Select proteome IDs at https://www.uniprot.org/proteomes/</span>
0047             <span class="comment">% or taxonomy IDs at https://www.uniprot.org/taxonomy.</span>
0048             obj.params.uniprot.type = <span class="string">'taxonomy'</span>; <span class="comment">% 'proteome' or 'taxonomy'</span>
0049             obj.params.uniprot.ID = <span class="string">'559292'</span>; <span class="comment">% should match the ID type</span>
0050             <span class="comment">% Field for Uniprot gene ID - should match the gene ids used in the</span>
0051             <span class="comment">% model. It should be one of the &quot;Returned Field&quot; entries under</span>
0052             <span class="comment">% &quot;Names &amp; Taxonomy&quot; at this page: https://www.uniprot.org/help/return_fields</span>
0053             obj.params.uniprot.geneIDfield = <span class="string">'gene_oln'</span>;
0054             <span class="comment">% Whether only reviewed data from UniProt should be considered.</span>
0055             <span class="comment">% Reviewed data has highest confidence, but coverage might be (very)</span>
0056             <span class="comment">% low for non-model organisms</span>
0057             obj.params.uniprot.reviewed = false;
0058 
0059             <span class="comment">% Reaction ID for glucose exchange reaction (or other preferred carbon source)</span>
0060             obj.params.c_source = <span class="string">'r_1714'</span>; 
0061 
0062             <span class="comment">% Reaction ID for biomass pseudoreaction</span>
0063             obj.params.bioRxn = <span class="string">'r_4041'</span>;
0064 
0065             <span class="comment">% Name of the compartment where the protein pseudometabolites</span>
0066             <span class="comment">% should be located (all be located in the same compartment,</span>
0067             <span class="comment">% this does not interfere with them catalyzing reactions in</span>
0068             <span class="comment">% different compartments). Typically, cytoplasm is chosen.</span>
0069             obj.params.enzyme_comp = <span class="string">'cytoplasm'</span>;
0070         <span class="keyword">end</span>
0071         
0072         <a name="_sub1" href="#_subfunctions" class="code">function [spont,spontRxnNames] = getSpontaneousReactions(obj,model)</a>
0073             <span class="comment">% Indicates how spontaneous reactions are identified. Here it</span>
0074             <span class="comment">% is done by the reaction have 'spontaneous' in its name.</span>
0075             spont = contains(model.rxnNames,<span class="string">'spontaneous'</span>);
0076             spontRxnNames = model.rxnNames(spont);
0077         <span class="keyword">end</span>
0078     <span class="keyword">end</span>
0079 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>